update profile registry doc (#174)
common-fate/docs: docs/granted/usage/profile-registry.md

You can add a repository to Granted's Profile Registry by running the following command:

```

> granted registry add -n <registry_name> -u <repository_url>

Reviewer: ```suggestion
granted registry add -name <registry_name> -u <repository_url>
```

# This section is automatically generated by Granted (https://granted.dev). Manual edits to this section will be overwritten.
# To edit, clone your profile registry repo, edit granted.yml, and push your changes. You may need to make a pull request depending on the repository settings.
# To stop syncing and remove this section, run 'granted registry remove
[granted_registry_start granted-registry]

[profile dev]
sso_start_url  = <https://example.awsapps.com/start>
sso_region     = <your-sso-region>
sso_account_id = <your-sso-account-id>
sso_role_name  = <your-sso-role-name>

[granted_registry_end granted-registry]
```

:::info
To avoid confusion, synced profiles are placed into a designated region in the `/.aws/config` file as shown above. This allows users to have a mix of both synced and regular profiles in their config file:
:::info

> ## Optional Flags when Adding a Profile Registry

Reviewer: ```suggestion
## Optional flags when adding a Profile Registry
```

| Optional Flag             | Alias | Description                                                                 |
|---------------------------|-------|-----------------------------------------------------------------------------|

> | path                      | p     | provide path if only the subfolder needs to be synced                       |
> | filename                  | f     | provide filename if yml file is not granted.yml                             |
> | priority                  |       | profile registry will be sorted by priority descending                      |
> | prefix-all-profiles       | pap   | provide this flag if you want to append registry name to all profiles       |
> | prefix-duplicate-profiles | pdp   | provide this flag if you want to append registry name to duplicate profiles |
> | requiredKey               | r     | used to bypass the prompt or override user specific values                  |

Reviewer: ```suggestion
| `-path`                      | `-p`     | provide path if only the subfolder needs to be synced                       |
| `-filename`                  | `-f`     | provide filename if yml file is not granted.yml                             |
| `-priority`                  |       | profile registry will be sorted by priority descending                      |
| `-prefix-all-profiles`       | `-pap`   | provide this flag if you want to append registry name to all `-profiles`       |
| `-prefix-duplicate-profiles` | `-pdp`   | provide this flag if you want to append registry name to duplicate profiles |
| `-requiredKey`               | `r`     | used to bypass the prompt or override user specific values                  |
```

Reviewer: @Eddie023 there needs to be some consistency across multi-word flags here. Currently we have things like 
- filename
- prefix-all-profiles
- requiredKey

Please choose one option ie fileName, prefixAllProfiles, requiredKey. I don't have a preference for which.

Then you can specify only the subfolder containing a `granted.yml` file:

```
granted registry add -n <name> -u <your-repo-url> -p <relative_path_to_subfolder>
```

Or with custom file name as:

```
granted registry add -n <name> -u <your-repo-url.git> -p <relative_path_to_subfolder> -f <filename>
```

> ### Deterministic Deduplication Priority 

Reviewer: ```suggestion
### Duplicate profile names
```

> When syncing multiple registries they are synced by priority descending. The default Priority is 0. 
> This lets users define which profile should have a registry name as prefix on profile duplication.

Reviewer: ```suggestion
When multiple registries are in use, registries are synced in descending priority where the default priority is zero. This allows you to define which profile registry is prefixed upon duplication. 
```

> For example if you have Profile Registries A and B with shared profile name `develop` then profile with lower pririty will be prefixed with registry name.

Reviewer: ```suggestion
For example, you have Profile Registries `core` and `devops` that both include a profile named `developer`. Profile Registry `core` is prefixed 0 and Profile Registry `devops`, 1. The result of this is the `developer` profile within the `devops` Profile Registry will be prefixed with the registry name `devops`.
```

> ## Variable Substitution in AWS profile 

Reviewer: ```suggestion
## Variable substitution for profile parameters
```

> If you want to template your AWS profile for user specific values or some re-useability then you can add `templateValues` in your config.yml file like:

Reviewer: ```suggestion
To enable templated profiles, profile values can include variable parameters. This is achieved through adding `templateValues` to your `config.yml` file. Below is an example:```

```
awsConfig:
  - ./dev/config1
templateValues:

>   - <YourKeyName>

Reviewer: ```suggestion
  - <your_key_name>
```

Author: @ellie-narducci  I didn't update this one because Go templates are generally added as PascalCase and these keys should be same as values in templates like `{{.Required.SessionName}}`

    - ... 
  - BaseProfile: 
    - value: "my-profile"
  - SessionName: 
      - isRequired: true 
      - prompt: "Enter your user name (e.g john.doe)"
```

:::info
You can use `{{ .Profile }}` when templating and will be used to replace with the correct profile name

> :::info

Reviewer: @Eddie023 I'm not quite sure I understand what this point is trying to achieve, could you provide some further context?

> ### User Specific Values 

Reviewer: ```suggestion
### User specific values 
```

> For cases, when there are user specific values that are required by users to be enter for Profile Registry. You can add `isRequired:true` field to the keyName to define them as User Specific Values.
> 
> For example: If you want user to enter `session_name` then 
> 
> 1. Add your template variable to config.yml file like:

Reviewer: ```suggestion
Should your Profile Registry require user specific values, you can set these values as required fields. You can achieve this by adding `isRequired:true` to the key name of the field. Below is an example `config.yml` file:

```

```
awsConfig:
  - ...
templateValues:
  - SessionName: 
      - isRequired: true 
      - prompt: "Enter your user name (e.g john.doe)"
```

> Your config file might look like:

Reviewer: ```suggestion
The above will output a profile that looks something like this:
```

```
[profile prod]
session_name = {{ .Required.SessionName }}
role_arn = arn:aws:iam::123456789012:role/example
```

When templating, for user specific values you need to add `{{ .Required.<KeyName>}}` 

> This will prompt user to add value for `SessionName` when granted registry add/sync command is run and will interpolate the template with the user provided value.

Reviewer: ```suggestion
The inclusion of `.Required` will prompt the user to add the value for `SessionName` when the `granted registry add/sync` command is run. Upon successful execution, Granted will interpolate the template to the AWS config file with the user provided value.
```

:::info

> You can bypass the prompt by providing `-r` flag to granted registry add like: <br/>

Reviewer: ```suggestion
You can bypass the prompt by providing the `-r` flag to the Granted CLI command:

```
granted registry add -n <name> -u <url> -r "SessionName=<your_session_name>"
```
```

> :::info
> You can bypass the prompt by providing `-r` flag to granted registry add like: <br/>
> `granted registry add -n <name> -u <url> -r "SessionName=JohnDoe"`
> :::info

Reviewer: ```suggestion
:::info
You can bypass the prompt by providing the `-r` flag to the Granted CLI command:
`granted registry add -n <name> -u <url> -r "SessionName=<your_session_name>"`

:::info
```

> ### Arbitary variables 

Reviewer: ```suggestion
### Arbitrary variables 
```

> For cases when you want to reuse a variable for partial config files. You can add `value` to any keyName in `templateValues`.
> 
> 1. Add your template variable to config.yml file like:

Reviewer: ```suggestion
For cases where you want to reuse a variable within a config file, you can add `value` to any key name in `templateValues`. Below is an example `config.yml` file:
```

```
awsConfig:
  - ...
templateValues:
  - BaseProfile: 
      - value: "my-profile"
```
 

> Your config file might look like:

Reviewer: ```suggestion
The above will output a profile that looks something like this:
```

This will loop over the repositories associated with your Profile Registry and will pull the latest changes from the remote origin, performing a sync operation.

:::info
When using AWS CLI with `granted credential-process` you will not be able to add user specific keys through STDIO.

> Instead the syncing will fail. This occurs as AWS expects only specific JSON STDOUT when configured with sourcing credentials with an external process. You cannot ask user to enter required keys. You should run `granted registry sync` to enter the required keys.

Reviewer: ```suggestion
When using AWS CLI with `granted credential-process` you will not be able to add user specific keys through STDIO. Doing so will cause the sync to fail. This occurs as AWS expects only specific JSON STDOUT when sourcing credentials via an external process. Instead, you should first run the following command to enter the required keys:

```bash
granted registry sync
```

```

This will display all repositories subscribed to Granted's Profile Registry and will prompt you to choose a repository to unsubscribe.

> ## Migrating Profile Registry Configuration

Reviewer: ```suggestion
## Migrating Profile Registry configuration
```

> Previously, Profile Registry didn't require passing `-n` or `-u` flag. However, the current version has some breaking changes so if you are already using Granted Profile Registry then run the following command to migrate your registry configuration.

Reviewer: ```suggestion
As part of [Granted v0.5.0](https://github.com/common-fate/granted/releases/tag/v0.5.0), the Profile Registry didn't require the `-n` or `-u` flags. [Granted v0.5.3](https://github.com/common-fate/granted/releases/tag/v0.5.3) includes breaking changes to this. If you are using a version of Profile Registries prior to Granted v0.5.3, run the following command to migrate your registry configuration:
```
